modules/dnn/src/darknet/darknet_io.cpp

```diff
@@ -0,0 +1,626 @@
+        return dst;
+    }
+
+    bool ReadDarknetFromCfgFile(const char *cfgFile, NetParameter *net)
+    {
+        std::ifstream ifile;
+        ifile.open(cfgFile);
+        if (ifile.is_open())
+        {
+            bool read_net = false;
+            int layers_counter = -1;
+            for (std::string line; std::getline(ifile, line);) {
+                if (line.empty()) continue;
+                switch (line[0]) {
+                case '\0': break;
+                case '#': break;
+                case ';': break;
+                case '[':
+                    if (line == "[net]") {
+                        read_net = true;
+                    }
+                    else {
+                        // read section
+                        read_net = false;
+                        ++layers_counter;
+                        std::string layer_type = line.substr(1, line.size() - 2);
+                        net->layers_cfg[layers_counter]["type"] = layer_type;
+                    }
+                    break;
+                default:
+                    // read entry
+                    size_t separator_index = line.find('=');
+                    if (separator_index != std::string::npos) {
+                        std::string name = line.substr(0, separator_index);
+                        std::string value = line.substr(separator_index + 1, line.size() - (separator_index + 1));
+                        name = escapeString(name);
+                        value = escapeString(value);
+                        if (name.empty() || value.empty()) continue;
+                        if (read_net)
+                            net->net_cfg[name] = value;
+                        else
+                            net->layers_cfg[layers_counter][name] = value;
+                    }
+                }
+            }
+
+            std::string anchors = net->layers_cfg[net->layers_cfg.size() - 1]["anchors"];
+            std::vector<float> vec = getNumbers<float>(anchors);
+            std::map<std::string, std::string> &net_params = net->net_cfg;
+            net->width = getParam(net_params, "width", 416);
+            net->height = getParam(net_params, "height", 416);
+            net->channels = getParam(net_params, "channels", 3);
+        }
+        else
+            return false;
+
+        layerShape current_shape;
+        current_shape.input_channels = net->channels;
+        current_shape.input_w = net->width;
+        current_shape.input_h = net->height;
+        net->out_shape_vec.resize(net->layers_cfg.size());
+
+        int layers_counter = -1;
+
+        setLayersParams setParams(net);
+
+        typedef std::map<int, std::map<std::string, std::string> >::iterator it_type;
+        for (it_type i = net->layers_cfg.begin(); i != net->layers_cfg.end(); ++i) {
+            ++layers_counter;
+            std::map<std::string, std::string> &layer_params = i->second;
+            std::string layer_type = layer_params["type"];
```

Comment:
Please, add an assertion for unknown layer type to prevent unexpected errors. In example, I can't read any model now because an every `layer_type` ends with `]` character (`convolutional]`, `maxpool]`).

Reply:
It works now but `Reproducibility_TinyYoloVoc` and `Reproducibility_YoloVoc` tests are failed for me. Do them passed locally?